### PR TITLE
[AEA-799] [AEA-908] Libp2p - Add multiple address lookup attempts to filter out outdated DHT records

### PR DESCRIPTION
**Pull request by lrahmani**
## Proposed changes

- Add multiple DHT agent address lookup until a registered provider is reachable
- Proper libp2p tests cleanup in case of assertion failure or setup error

### CODE REVIEW

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go [hunk 1]
```diff
@@ -526,8 +526,8 @@ func (dhtPeer *DHTPeer) RouteEnvelope(envel *aea.Envelope) error {
 }
 
 func (dhtPeer *DHTPeer) lookupAddressDHT(address string) (peer.ID, error) {
-	lerror, _, linfo, _ := dhtPeer.getLoggers()
-
+	lerror, lwarn, linfo, _ := dhtPeer.getLoggers()
+	var err error
 	addressCID, err := utils.ComputeCID(address)
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:531]: err will be overwritten in 531, intended?

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go [hunk 2]
```diff
@@ -537,35 +537,54 @@ func (dhtPeer *DHTPeer) lookupAddressDHT(address string) (peer.ID, error) {
...
+				break
+			}
+
+			dhtPeer.routedHost.Peerstore().ClearAddrs(provider.ID)
+			lwarn().Str("op", "lookup").Str("addr", address).Msgf("couldn't open stream to address provider %s: %s", provider, err.Error())
+			lwarn().Str("op", "lookup").Str("addr", address).Msgf("looking up for other providers...")
+		}
+
+		if provider.ID == "" {
+			msg := "didn't found any provider for address"
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:574]: didn't found > didn't find

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go [hunk 3]
```diff
@@ -537,35 +537,54 @@ func (dhtPeer *DHTPeer) lookupAddressDHT(address string) (peer.ID, error) {
...
-		err = errors.New("didn't found any provider for address within timeout")
-		lerror(err).Str("op", "lookup").Str("addr", address).Msg("")
-		select {
-		default:
-			time.Sleep(200 * time.Millisecond)
-			providers = dhtPeer.dht.FindProvidersAsync(ctx, addressCID, 1)
-			provider = <-providers
+
+	for !connected {
+		providers := dhtPeer.dht.FindProvidersAsync(ctx, addressCID, 0)
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:548]: previously the third arg was 1, intentional?

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go [hunk 4]
```diff
@@ -537,35 +537,54 @@ func (dhtPeer *DHTPeer) lookupAddressDHT(address string) (peer.ID, error) {
 		Msgf("Querying for providers for cid %s...", addressCID.String())
 	ctx, cancel := context.WithTimeout(context.Background(), addressLookupTimeout)
 	defer cancel()
-	providers := dhtPeer.dht.FindProvidersAsync(ctx, addressCID, 1)
+	var elapsed time.Duration
+	var provider peer.AddrInfo
+	var connected bool = false
+	var s network.Stream
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:543]: single letter variables are hard to understand for new devs; call it stream?

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go (hunk 3 above)
Comment (lrahmani) [reply packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:548]: yes

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go (hunk 4 above)
Comment (lrahmani) [reply packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:543]: sure

File: tests/test_packages/test_connections/test_p2p_libp2p/test_fault_tolerance.py [hunk 5]
```diff
@@ -155,10 +164,133 @@ def test_envelope_routed_after_relay_restart(self):
...
+            message_id=1,
+            target=0,
+            performative=DefaultMessage.Performative.BYTES,
+            content=b"helloAfterChangingPeer",
+        )
+        envelope = Envelope(
+            to=addr_2,
+            sender=addr_1,
+            protocol_id=DefaultMessage.protocol_id,
+            message=DefaultSerializer().encode(msg),
```
Comment (DavidMinarsch) [inline tests/test_packages/test_connections/test_p2p_libp2p/test_fault_tolerance.py:276]: FYI; you can do `msg.protocol_id` and `msg.encode()`

**Review by DavidMinarsch — APPROVED** [review]
LGTM, some minor changes requested

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go (hunk 3 above)
Comment (DavidMinarsch) [reply packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:548]: Would help to add a comment what the 0 is or Constantize it..another developer will be left guessing

File: packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go (hunk 1 above)
Comment (lrahmani) [reply packages/fetchai/connections/p2p_libp2p/dht/dhtpeer/dhtpeer.go:531]: yes, it is

**Review by DavidMinarsch — APPROVED** [review]

**Pull request merged**